solutions/POI 05-sza.mdx

```diff
@@ -0,0 +1,218 @@
+
+</Spoiler>
+
+</Spoiler>
+
+<Spoiler title="Hint 3">
+
+First, use KMP to compute the prefix function $\pi$ of $S$. The shortest template of $S$ is either $S$ itself or the shortest template of $S[0 : \pi[|S| - 1]]$.
+
+<Spoiler title="Proof">
+    
+We'll prove instead that any template of $S$ that isn't $S$ itself is also a template of $S[0 : \pi[|S| - 1]]$, since that's equivalent.
+
+Since $S[0 : \pi[|S| - 1]]$ is the longest proper prefix of $S$ that is also a suffix, we know that the only longer string that can be $S$'s template is $S$ itself. Let the shortest template of $S[0 : \pi[|S| - 1]]$ be $T'$ and some longer template of $S$ be $T$.
+
+By hint 1, $T$ must be a prefix and suffix of $S$, which means that it's also a prefix and suffix of $S[0 : \pi[|S| - 1]]$.
+
+This means that we can make at least half of $T$'s prefix using $T'$, since we can make $S[0 : \pi[|S| - 1]]$ (which is bigger than $T$) using $T'$ and $|T'| < |T|$. A similar argument holds for being able to make at least half of $T$'s suffix.
+
+This implies that $T'$ is a template of $T$, as we can combine this prefix and suffix to form the whole string. Therefore $T$ is also a template of $S[0 : \pi[|S| - 1]]$.
+
+</Spoiler>
+
+</Spoiler>
+
+<Spoiler title="Hint 4">
+
+Don't be afraid to use KMP again!
+
+</Spoiler>
+
+<Spoiler title="Solution 1">
+
+**Time complexity:** $\mathcal O(N \log N)$.
+
+First, we compute the prefix function $\pi$ of $S$ using KMP.
+
+If $2\pi[i - 1] \geq i$, then $S[0 : i]$ will never be the shortest template of $S$ because we know that $S[0 : \pi[i - 1]]$ is a template of $S[0 : i]$.
+
+This means that we only have $\mathcal O(\log N)$ candidate templates to check! We can then use KMP again on each candidate and check whether the maximum "gap" between matches is sufficiently small.
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+string s;
+int p[500000];
+
+int shortest_template(int n) {
+    if (!p[n - 1]) return n;
+    if (p[n - 1] * 2 >= n) return shortest_template(p[n - 1]);
+    int cand = shortest_template(p[n - 1]);
+    for (int i = 0, j = 0, curr = -1; i < n; i++) {
+        while (j && (j == cand || s[i] != s[j])) j = p[j - 1];
+        if (s[i] == s[j]) j++;
+        if (j == cand) curr = i;
+        if (i - curr >= cand) return n;
+    }
+    return cand;
+}
+
+int main() {
+    cin.tie(0)->sync_with_stdio(0);
+    cin >> s;
+    int n = s.size();
+    for (int i = 1; i < n; i++) {
+        p[i] = p[i - 1];
+        while (p[i] && s[i] != s[p[i]]) p[i] = p[p[i] - 1];
+        if (s[i] == s[p[i]]) p[i]++;
+    }
+    cout << shortest_template(n);
+    return 0;
+}
+```
+
+</Spoiler>
+
+<Spoiler title="Solution 2">
+
+**Time complexity:** $\mathcal O(N \log N)$.
+
+<Warning>
+
+The official time limits are very tight and `std::set`/`std::multiset` have very bad constant factors, so this code doesn't actually get 100 points. Nevertheless, we think the idea behind this solution is interesting enough to share.
```

Comment:
You can make this O(N) by starting with `matches` containing all indices and then erasing elements one by one. Use a linked list instead of a set.

Reply:
Rip my O(N) still gets TLE because of `vector` and the strict TL: https://github.com/dolphingarlic/CompetitiveProgramming/blob/master/POI/POI%2005-template/O(N)%20KMP%20tree.cpp (it times out while computing the prefix function)